### PR TITLE
HW05

### PR DESCRIPTION
**Pull request by psmishenko**

### CODE REVIEW

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+require_relative '../pavel_mishchenko_hw_03_t_01'
+
+RSpec.describe Homework3 do
+  subject { described_class.new.task1(log) }
+
+  describe 'Log that has right output' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:9]: it's better to use here method name

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    context 'when logs have only errors' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:10]: what about the context when logs have no errors?

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_02_spec.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+require_relative '../pavel_mishchenko_hw_03_t_02'
+
+RSpec.describe Homework3 do
+  subject { described_class.new.task2(log) }
+
+  describe 'Log that has wrong output' do
+    context 'when log is not String class' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_02_spec.rb:10]: did you forget about other contexts?

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+require_relative '../pavel_mishchenko_hw_03_t_03'
+
+RSpec.describe Homework3 do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb:6]: all previous comments apply to this file as well

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb [hunk 5]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+require_relative '../pavel_mishchenko_hw_03_t_04'
+
+RSpec.describe Homework3 do
+  subject { described_class.new.task4(str) }
+
+  describe 'counter of letters and digits in the string' do
+    context 'when word consists letters and digits' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb:10]:
```suggestion
    context 'when string consists of letters and digits' do
```

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  describe '#notify_student' do
+    context "when method hasn't arguments" do
+      subject { mentor.notify_student }
+
+      it 'raises ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when mentor notifies student' do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:20]: the context should be with opposite meaning

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+      subject do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:21]: try not create subject each time ( applies to all identical moments in the code )

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  let(:mentor) { described_class.new(name: 'Test', surname: 'Testov') }
+  let(:notification1) { Notification.new(title: 'TestTitle1', description: 'TestDescription1') }
+  let(:notification2) { Notification.new(title: 'TestTitle2', description: 'TestDescription2') }
+
+  describe '#notify_student' do
+    context "when method hasn't arguments" do
+      subject { mentor.notify_student }
+
+      it 'raises ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:16]: as you use subject, you can read about [such construction](https://relishapp.com/rspec/rspec-core/docs/subject/one-liner-syntax)

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb (hunk 6 above)
Comment (DashaSledopit) [reply HW04/Pavel_Mishchenko/spec/mentor_spec.rb:20]: you can test here that notifications_list includes the necessary description much easier 
try not to use such huge subject and the matcher `an_object_having_attributes`

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+require_relative 'spec_helper'
+
+RSpec.describe Mentor do
+  let(:student) { Student.new(name: 'TestName', surname: 'TestSurname') }
+  let(:mentor) { described_class.new(name: 'Test', surname: 'Testov') }
+  let(:notification1) { Notification.new(title: 'TestTitle1', description: 'TestDescription1') }
+  let(:notification2) { Notification.new(title: 'TestTitle2', description: 'TestDescription2') }
+
+  describe '#notify_student' do
+    context "when method hasn't arguments" do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:12]:
```suggestion
    context "when method has no arguments" do
```

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+
+    context 'when homework is not Homework.class instance' do
+      let(:homework) { 404 }
+
+      it 'raises NoMethodError' do
+        expect { subject }.to raise_error(NoMethodError)
+      end
+    end
+
+    context 'when mentor reject homework' do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:116]: the same thing: try use the opposite contexts
for instance, `when arguments are valid`

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+      subject { student.to_work! }
+
+      it 'raises ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when student takes to work his homework' do
+      it 'changes notification list of mentor' do
+        expect { subject }.to change(mentor, :notifications_list).from([])
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/student_spec.rb:88]: if you add `from` you can also add `to` :)  everywhere where you use `change`

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 12]
```diff
@@ -0,0 +1,251 @@
...
+      Homework.new(title: 'TestTitle',
+                   description: 'TestDescription',
+                   student: student,
+                   mentor: mentor)
+    end
+
+    context 'when student has homeworks' do
+      before do
+        student.homeworks_list.push(homework)
+      end
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/student_spec.rb:187]:
```suggestion
      before { student.homeworks_list.push(homework) }
```

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb (hunk 8 above)
Comment (psmishenko) [reply HW04/Pavel_Mishchenko/spec/mentor_spec.rb:16]: The one-liner syntax only works with non-block expectations (e.g. expect(obj).to eq, etc) and it cannot be used with block expectations (e.g. expect { object }).

Comment (DashaSledopit) [reply HW04/Pavel_Mishchenko/spec/mentor_spec.rb:16]: but there are other situations, where is_expected can be used

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+            10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] "POST /grid/2/event HTTP/1.1" 200 - 0.2277
+            2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>
+            10.6.246.101 - - [23/Apr/2018:20:29:39 +0300] "POST /grid/2/messages HTTP/1.1" 200 48 0.0290
+          LOGS
+        end
+        let(:right_output) do
+          '2018-04-23 20:30:42: SSL error, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSLError: System error: Undefined error: 0 - 0>'
+        end
+
+        it('returns full text of the first line with an error') { is_expected.to eq(right_output) }
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:24]:
```suggestion
        it { is_expected.to eq(right_output) }
```

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    subject { described_class.new.task4(str) }
+
+    context 'when argument is valid' do
+      context 'when string is empty' do
+        let(:str) { '' }
+
+        it('returns hash with 0 values') { is_expected.to eq({ digits: 0, letters: 0 }) }
+      end
+
+      context 'when sring consists of letters and digits' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb:17]:
```suggestion
      context 'when string consists of letters and digits' do
```

File: HW04/Pavel_Mishchenko/spec/student_spec.rb (hunk 11 above)
Comment (DashaSledopit) [reply HW04/Pavel_Mishchenko/spec/student_spec.rb]: please, add it everywhere where you use change

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+        expect { subject }.to raise_error(NoMethodError)
+      end
+    end
+  end
+
+  describe '#mark_as_read!' do
+    context 'when notifiications is unread' do
+      before do
+        mentor.notifications_list.push(notification1)
+        mentor.notifications_list.push(notification2)
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:62]: could you do it in 1 line?

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 16]
```diff
@@ -0,0 +1,206 @@
...
+      end
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:63]: could you do it in 1 line?

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 17]
```diff
@@ -0,0 +1,206 @@
...
+  describe '#notify_student' do
+    context 'when no arguments given' do
+      subject { mentor.notify_student }
+
+      it 'raises ArgumentError' do
+        expect { subject }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when arguments are valid' do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:20]: Could you remake this context according to the example of the method #mark_as_read! ?

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 18]
```diff
@@ -0,0 +1,206 @@
...
+      end
+    end
+
+    context 'when notifiications is read' do
+      subject { mentor.mark_as_read! }
+
+      before do
+        notification1.status, notification2.status = 'read'
+        mentor.notifications_list.push(notification1)
+        mentor.notifications_list.push(notification2)
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:77]: it can be done in 1 line

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 19]
```diff
@@ -0,0 +1,206 @@
...
+  end
+
+  describe '#notifications' do
+    subject { mentor.notifications }
+
+    context 'when mentor has only unread notifications' do
+      before do
+        mentor.notifications_list.push(notification1)
+        mentor.notifications_list.push(notification2)
+      end
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:163]: It also can be done in 1 line

Comment (DashaSledopit) [reply HW04/Pavel_Mishchenko/spec/mentor_spec.rb:163]: and other similar situations

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 20]
```diff
@@ -0,0 +1,239 @@
+# frozen_string_literal: true
+
+require_relative 'spec_helper'
+
+RSpec.describe Student do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/student_spec.rb:5]: Please, add all the changes from the comments for the previous class to this class as well

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb (hunk 13 above)
Comment (psmishenko) [reply HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb]: Looks weird in console, but ok.

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 21]
```diff
@@ -0,0 +1,230 @@
...
+    end
+  end
+
+  describe '#add_answer!' do
+    subject do
+      student.add_answer!(homework, 'TestAnswer')
+      homework
+    end
+
+    let(:homework) do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/student_spec.rb:92]: you use homework multiple times, maybe it's a good idea to move it to the beginning of the class?

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+  describe '#task3' do
+    subject { described_class.new.task3(log) }
+
+    context 'when log that has right output' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb:10]: maybe it's better to use contexts : 'when log is valid' and the opposite one? in all such situations

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 23]
```diff
@@ -0,0 +1,170 @@
...
+    end
+
+    context 'when student is not Student.class instance' do
+      let(:student) { 404 }
+
+      it 'raises NoMethodError' do
+        expect { subject }.to raise_error(NoMethodError)
+      end
+    end
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:54]: Unnecessary

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+  end
+
+  describe '#mark_as_read!' do
+    context 'when notifiications is unread' do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:58]:
```suggestion
    context 'when notifiication is unread' do
```

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+      before { mentor.notifications_list.push(notification) }
+
+      it 'makes all notifictions as read' do
+        mentor.mark_as_read!
+        expect(mentor.notifications_list).to all(have_attributes(status: 'read'))
+      end
+    end
+
+    context 'when notifiications is read' do
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:67]:
```suggestion
    context 'when notifiication is read' do
```

File: HW04/Pavel_Mishchenko/spec/mentor_spec.rb [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+      end
+    end
+
+    context 'when homework is not Homework.class instance' do
+      let(:homework) { 404 }
+
+      it 'raises ArgumentError' do
+        expect { subject }.to raise_error(NoMethodError)
+      end
+    end
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/mentor_spec.rb:118]: Unnecessary

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 27]
```diff
@@ -0,0 +1,189 @@
...
+  describe '#to_work!' do
+    subject { student.to_work!(homework) }
+
+    context 'when homework is not Homework.class instance' do
+      let(:homework) { 404 }
+
+      it 'raises NoMethodError' do
+        expect { subject }.to raise_error(NoMethodError)
+      end
+    end
```
[the remark this answers is not in the document]
Comment (DashaSledopit) [reply HW04/Pavel_Mishchenko/spec/student_spec.rb:62]: in all similar cases

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+            10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] "POST /test/2/messages HTTP/1.1" 200 48 0.0290
+          LOGS
+        end
+
+        it { is_expected.to eq('') }
+      end
+    end
+
+    context 'when log that has wrong output' do
+      context 'when text are not given' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:75]:
```suggestion
      context 'when log is empty' do
```
```suggestion








      context 'when text are not given' do
```

Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:75]:
```suggestion
      context 'when log is empty' do
```
```suggestion








      context 'when text are not given' do
```

Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:75]:
```suggestion
      context 'when log is empty' do
```
```suggestion
      context 'when text are not given' do
```

Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:75]:
```suggestion
      context 'when log is empty' do
```
```suggestion
      context 'when text are not given' do
```

Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_01_spec.rb:75]:
```suggestion
      context 'when log is empty' do
```

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_02_spec.rb [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+
+require_relative 'spec_helper'
+require_relative '../pavel_mishchenko_hw_03_t_02'
+
+RSpec.describe Homework3 do
+  describe '#task2' do
+    subject { described_class.new.task2(log) }
+
+    context 'when log that has wrong output' do
+      context 'when text are not given' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_02_spec.rb:11]:
```suggestion
      context 'when log is empty' do
```

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb [hunk 30]
```diff
@@ -0,0 +1,109 @@
...
+            2018-04-23 17:18:19.5 ubuntu-xenial[14319] Debug - monitoring grid communication health
+            2018-04-23 17:18:38.8 ubuntu-xenial[14319] Debug - connecting to: 10.6.246.101
+            2018-04-23 17:18:59.8 ubuntu-xenial[14319] Debug - inside docker_handle_event
+          LOGS
+        end
+
+        it { is_expected.to eq('0') }
+      end
+
+      context 'when there is two valid event' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_03_spec.rb:57]:
```suggestion
      context 'when there is two valid events' do
```

File: HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+
+      context 'when string consists of letters,symbols and digits' do
+        let(:str) { '01122333qwe_-+=}{()?<>' }
+
+        it { is_expected.to eq({ digits: 8, letters: 3 }) }
+      end
+    end
+
+    context 'when argument is invalid' do
+      context 'when no arguments is given' do
```
Comment (DashaSledopit) [inline HW03/pavel_mishchenko/spec/pavel_mishchenko_hw_03_t_04_spec.rb:31]:
```suggestion
      context 'when no arguments are given' do
```

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 32]
```diff
@@ -0,0 +1,189 @@
...
+
+    context 'when no argument given' do
+      it 'raises ArgumentError' do
+        expect { student.to_check! }.to raise_error(ArgumentError)
+      end
+    end
+
+    context 'when argument is valid' do
+      it 'changes notification list of mentor' do
+        expect { subject }.to change(mentor, :notifications_list)
```
[the remark this answers is not in the document]
Comment (psmishenko) [reply HW04/Pavel_Mishchenko/spec/student_spec.rb]: Sorry, idk how I miss it

File: HW04/Pavel_Mishchenko/spec/student_spec.rb [hunk 33]
```diff
@@ -0,0 +1,163 @@
...
+      it 'changes notification list of mentor' do
+        expect { subject }.to change(mentor, :notifications_list).from([]).to([Notification])
+      end
+    end
+  end
+
+  describe '#add_answer!' do
+    subject do
+      student.add_answer!(homework, 'TestAnswer')
+      homework
```
Comment (DashaSledopit) [inline HW04/Pavel_Mishchenko/spec/student_spec.rb:72]: why do we need this?

**Pull request merged**